### PR TITLE
Support caching for @jitclass

### LINKED ISSUE #6522: Caching a function that takes in a jitclass instance
```python
@jitclass([])
class NumbaClass():
    def __init__(self):
        pass

@njit(cache=True)
def calc(numbaObj):
    pass

numbaObject = NumbaClass()
calc(numbaObject)
```

Each time the script above is ran, it generates a new cache entry for "calc", likely because NumbaClass.class_type.instance_type changes on every run, making the cache flag effectively not work and growing the \_\_pycache__ folder on every run (mine eneded up being a few gigs). Is there a work-around for this? Numba 0.51.2

### PR DESCRIPTION
**Pull request by HIEROT**
To fix #6522, as one comment pointed out, there is no substantial difficulty in caching a jitted class and functions taking them as params or returns, and it will be very useful to achieve it. Since the isuue has been unresolved for long, I made some efforts to do it with minimum code modification.

Now I will explain why original codes failed to cache a jitted class and what I have done. I noticed that the consideration of cache collision prompted the author to hardcoded addresses, but the ability to cache them is more important and there are other ways to solve it.

- The name of ClassType and DeferredType hardcodes type addresses. I see no point in it so they are removed as in core.types.
- On boxing at function return, as in experimental.jitclass.boxing, a dynamically created _specialize_box subtype's address is linked in the module. However we can call pythonapi with constant string param to access these methods, just like praactices in typeddict or typedlist. On the other hand we need to create this _specialize_box elsewhere as it will not be called again if cache is loaded as where it is now. So I put it in @jitclass decorator
- If a dynamically jitted jitclass (as it's useful in dtype generic) constructor is passed as a closure in a jitted function, since there are addresses in its state, naive pickling will result in a cache miss, I (temporarily) make it scan for user defined \_identity\_ attribute, if it exists, then hash it to determine cache key. The idea is that the user should determine if their code is oudated.
- Stamps of cached functions are now based on source code stripped of docstring and comments so it will not recompile at modifying other parts in the source file.

After these changes:
- A jitted class instance can be a param or return value.
- A dynamically jitted classes' constructor can be cached as a closure, as long as you jit them every time before any cached function using it is called.
- methods and methods overloaded elsewhere are cachable too (Becareful when modifying your code as they may silently obsolete other cached functions using them!) (no including property getter or setters as they are too simple to cache)

Limitations:
- Now the cache expire detection is in no way as restrict as before but I think there are solutions other than hardcoded address to solve it, and we can develop it in the future. For now I am using this in my research code.

P.S.
- This PR is not meticulously polished to industrial standard but is now promoted for anyone in eager requirement. It needs further development for jitclass really useful.
- Basically @jitclass is a fancy wrap of structref, so it should be easy to cache them.

### CODE REVIEW

File: numba/experimental/jitclass/boxing.py [hunk 1]
```diff
@@ -203,18 +203,15 @@ def _box_class_instance(typ, val, c):
...
-    )
-    box_cls = c.builder.bitcast(voidptr_boxcls, c.pyapi.pyobj)
-
+    modname = c.context.insert_const_string(
+        c.builder.module, 'numba.experimental.jitclass.boxing')
+    jitclass_boxing_mod = c.pyapi.import_module_noblock(modname)
+    box_subclass_cache = c.pyapi.object_getattr_string(
+        jitclass_boxing_mod, '_cache_specialized_box')
+    box_cls = c.pyapi.dict_getitem_string(box_subclass_cache, typ.name)
     box = c.pyapi.call_function_objargs(box_cls, ())
```
Comment (HIEROT) [inline numba/experimental/jitclass/boxing.py:212]: Maybe some decref is needed here.
